crates/bevy_asset/src/io/wasm/opfs.rs

```diff
@@ -0,0 +1,735 @@
+                std::io::ErrorKind::NotFound,
+            ))
+            .map(|value| value.unchecked_into())?;
+    }
+
+    Ok(current)
+}
+
+/// Get child entries of this directory.
+async fn get_entries(start: &FileSystemDirectoryHandle) -> impl Stream<Item = PathBuf> + Unpin {
+    JsStream::from(start.keys())
+        .flat_map(|result| futures_lite::stream::iter(result.ok()))
+        .flat_map(|value| futures_lite::stream::iter(value.dyn_into::<JsString>().ok()))
+        .map(String::from)
+        .map(PathBuf::from)
+}
+
+/// Open a file relative to `start` from a given `path`.
+/// Will create directories and the final file based on the provided `path` if `create` is `true`.
+async fn get_file(
+    start: &FileSystemDirectoryHandle,
+    path: impl IntoIterator<Item = &str>,
+    create_file: bool,
+    create_path: bool,
+) -> std::io::Result<FileSystemFileHandle> {
+    let mut current = start.clone();
+
+    let mut iter = path.into_iter().peekable();
+
+    let options = {
+        let mut options = FileSystemGetDirectoryOptions::new();
+        options.create(create_path);
+        options
+    };
+
+    let file_name = loop {
+        let Some(path) = iter.next() else {
+            break None;
+        };
+
+        if iter.peek().is_none() {
+            break Some(path);
+        };
+
+        current = current
+            .get_directory_handle_with_options(path, &options)
+            .into_js_future()
+            .await
+            .map_err(js_value_to_err(
+                "Cannot get Directory Handle",
+                std::io::ErrorKind::NotFound,
+            ))
+            .map(|value| value.unchecked_into())?;
+    };
+
+    let Some(file_name) = file_name else {
+        return Err(std::io::Error::new(
+            std::io::ErrorKind::InvalidInput,
+            "Provided path is empty",
+        ));
+    };
+
+    current
+        .get_file_handle_with_options(file_name, &{
+            let mut options = FileSystemGetFileOptions::new();
+            options.create(create_file);
+            options
+        })
+        .into_js_future()
+        .await
+        .map_err(js_value_to_err(
+            "File not available",
+            std::io::ErrorKind::NotFound,
+        ))
+        .map(|value| value.unchecked_into())
+}
+
+async fn remove_entry(handle: &FileSystemDirectoryHandle, entry: &str) -> std::io::Result<()> {
+    handle
+        .remove_entry_with_options(entry, &{
+            let mut options = FileSystemRemoveOptions::new();
+            options.recursive(true);
+            options
+        })
+        .into_js_future()
+        .await
+        .map_err(js_value_to_err(
+            "Cannot remove Directory",
+            std::io::ErrorKind::Other,
+        ))?
+        .is_undefined()
+        .then_some(())
+        .ok_or(std::io::Error::new(
+            std::io::ErrorKind::Other,
+            "Failed to remove entry",
+        ))
+}
+
+mod utils {
```

Comment:
There's a lot of relatively generic WASM-async bridging code here in an inline submodule of `opfs` - it rubs me the wrong way a little.

Reply:
I agree. Part of the reason why I've moved so much into here is to help split out the more generic code and (hopefully) replace it or give it a better home (maybe `bevy_utils` with a WASM section?). I am hoping to remove at least _some_ of these.